contributing/writing-carbon-code.md

```diff
@@ -0,0 +1,158 @@
+# Writing Carbon code
+
+## Contents
+
+- [Introduction](#introduction)
+- [Writing components](#writing-components)
+- [Documenting with Storybook](#documenting-with-storybook)
+- [Linting](#linting)
+
+## Introduction
+
+This guide outlines some common practices for those contributing to Carbon to ensure consistency for consumers and other developers also working on the library.
+
+## Writing components
+
+### Folder Structure
+
+Our components should follow a consistent file structure as shown below (note the component names here are arbitrary).
+
+```none
+src/components/textbox/
+  |-- textbox.component.(ts|tsx)
+  |-- textbox.style.(ts|tsx)
+  |-- textbox.spec.(ts|tsx)
+  |-- textbox.stories.(ts|tsx)
+  |-- textbox.stories.mdx
+  |-- index.ts
+  |-- sub-component/
+      |-- ...
+  |-- __internal__/
+      |-- ...
+```
+
+Any sub-component that will not be exported directly from Carbon should go in the `__internal__/` folder. Sub-components should have the same file structure as root directory of the component.
+
+The root directory of the component should also have a `index.ts` file which exports the component, non-internal sub-components and their prop interfaces.
+
+### Component interface
+
+All component prop interfaces (in `<component-name>.component.(ts|tsx)` files) should follow this template:
+
+```ts
+// in textbox.component.tsx
+
+import React from "react";
+
+export interface TextboxProps {
+  /** Example required prop description */
+  exampleRequiredProp: string;
+  /** Example optional prop description */
+  exampleOptionalProp?: string;
+}
+
+Textbox.displayName = "Textbox";
+export default Textbox;
+```
+
+### Index file
+
+Exports in the `index.ts` file should follow this template:
+
+```ts
+// If component is sole export of *.component file, export it as default
+export { default } from "./textbox.component";
+export { type TextboxProps } from "./textbox.component"; 
+
+// If component should come with subcomponents - export all as named exports
+export { default as Select } from "./select.component";
+export { type SelectProps } from "./select.component";
+
+export { default as Option } from "./option/option.component";
+export { type OptionProps } from "./option/option.component";
+```
+
+> If you find any inconsistencies in module structure you are not obligated to change it. For example, if you need to work on that component to fix an unrelated bug. However, we are committed to provide a consistent interface for our components and will make changes where appropriate.
```

Comment:
May want to include that any changes will be appropriately semantically versioned